zkevm-circuits/src/evm_circuit/execution/create.rs

```diff
@@ -0,0 +1,915 @@
+        // Use rw_counter of the step which triggers next call as its call_id.
+        let callee_call_id = cb.curr.state.rw_counter.clone();
+        let current_call_id = cb.curr.state.call_id.clone();
+        let is_success = cb.query_bool();
+
+        // read from call context
+        let tx_id = cb.call_context(None, CallContextFieldTag::TxId);
+        let depth = cb.call_context(None, CallContextFieldTag::Depth);
+        let mut reversion_info = cb.reversion_info_read(None);
+
+        let keccak_output = cb.query_word_rlc();
+        let create = ContractCreateGadget::construct(cb);
+        let contract_addr = expr_from_bytes(&keccak_output.cells[..N_BYTES_ACCOUNT_ADDRESS]);
+        let contract_addr_rlc = cb.word_rlc::<N_BYTES_ACCOUNT_ADDRESS>(
+            keccak_output
+                .cells
+                .iter()
+                .take(N_BYTES_ACCOUNT_ADDRESS)
+                .map(Expr::expr)
+                .collect::<Vec<_>>()
+                .try_into()
+                .unwrap(),
+        );
+
+        // stack operations
+        let value = cb.query_word_rlc();
+        let offset = cb.query_cell_phase2();
+        let length = cb.query_word_rlc();
+        cb.stack_pop(value.expr());
+        cb.stack_pop(offset.expr());
+        cb.stack_pop(length.expr());
+        cb.condition(is_create2.expr(), |cb| {
+            cb.stack_pop(create.salt_word_rlc(cb).expr());
+        });
+        cb.stack_push(is_success.expr() * contract_addr_rlc.clone());
+
+        // read caller's balance and nonce
+        let caller_nonce = create.caller_nonce();
+        let caller_balance = cb.query_word_rlc();
+        cb.account_read(
+            create.caller_address(),
+            AccountFieldTag::Balance,
+            caller_balance.expr(),
+        );
+        cb.account_read(
+            create.caller_address(),
+            AccountFieldTag::Nonce,
+            caller_nonce.expr(),
+        );
+
+        // Pre-check: call depth, user's nonce and user's balance
+        let is_depth_in_range = LtGadget::construct(cb, depth.expr(), 1025.expr());
+        let is_insufficient_balance = LtWordGadget::construct(cb, &caller_balance, &value);
+        let is_nonce_in_range = LtGadget::construct(cb, caller_nonce.expr(), u64::MAX.expr());
+        let is_precheck_ok = and::expr([
+            is_depth_in_range.expr(),
+            not::expr(is_insufficient_balance.expr()),
+            is_nonce_in_range.expr(),
+        ]);
+
+        // verify gas cost
+        let init_code = MemoryAddressGadget::construct(cb, offset, length);
+        let memory_expansion = MemoryExpansionGadget::construct(cb, [init_code.address()]);
+        let init_code_word_size = ConstantDivisionGadget::construct(
+            cb,
+            init_code.length() + (N_BYTES_WORD - 1).expr(),
+            N_BYTES_WORD as u64,
+        );
+        let keccak_gas_cost =
+            GasCost::COPY_SHA3.expr() * is_create2.expr() * init_code_word_size.quotient();
```

Comment:
Since we have converted to use [shanghai geth traces](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1424), reference geth [gasCreateEip3860](https://github.com/ethereum/go-ethereum/blob/eb83e7c54021573eaceb14236af3a7a8c64f6027/core/vm/gas_table.go#L305) and [gasCreate2Eip3860](https://github.com/ethereum/go-ethereum/blob/eb83e7c54021573eaceb14236af3a7a8c64f6027/core/vm/gas_table.go#LL321C6-L321C23), it may be fixed to:
```
let keccak_gas_cost = init_code_word_size.quotient()
    * select::expr(
        is_create2.expr(),
        eth_types::evm_types::INIT_CODE_WORD_GAS,
        (INIT_CODE_WORD_GAS + GasCost::COPY_SHA3).expr(),
    );
```

Reply:
fixed in [7aa2d78](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1430/commits/7aa2d78031f1bf7f2e7831d27500410a94d30468)